src/ray/object_manager/pull_manager.cc

```diff
@@ -314,60 +314,44 @@ void PullManager::OnLocationChange(const ObjectID &object_id,
 }
 
 void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
+  // The object is already local; abort.
   if (object_is_local_(object_id)) {
     return;
   }
+
+  // The object is no longer needed; abort.
   if (active_object_pull_requests_.count(object_id) == 0) {
     return;
   }
+
+  // The object waiting for local pull retry; abort.
   auto it = object_pull_requests_.find(object_id);
   RAY_CHECK(it != object_pull_requests_.end());
   auto &request = it->second;
   if (request.next_pull_time > get_time_()) {
     return;
   }
 
-  // We always pull objects from a remote node before
-  // restoring it because of two reasons.
-  // 1. This will help reducing the load of external storages
-  //    or remote node that spilled the object.
-  // 2. Also, if we use multi-node file spilling, the restoration will be
-  //    confirmed by a object location subscription, so we should pull first
-  //    before requesting for object restoration.
-  bool did_pull = PullFromRandomLocation(object_id);
-  if (did_pull) {
-    // New object locations were found, so begin trying to pull from a
-    // client.
+  // If we can restore directly from this raylet, then prefer to do so.
+  bool can_restore_directly =
+      request.spilled_url.empty() &&
+      (request.spilled_node_id.IsNil() || request.spilled_node_id == self_node_id_);
```

Comment:
I assume that we'd want to pull from a remote node's object store instead of restoring an object from external storage, if possible, so shouldn't this contain an empty object locations check?

Reply:
Let me fix this.

Reply:
Why the check for `request.spilled_node_id.IsNil()`? Wouldn't that mean the node failed?

Reply:
A URL with node_id Nil means the request was spilled to remote storage (e.g., S3).